core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -337,95 +351,80 @@ public Builder removeProperties(Set<String> propertiesToRemove) {
-      ImmutableMap.Builder<Integer, Schema> builder = ImmutableMap.builder();
-      for (Schema schema : schemasToIndex) {
-        builder.put(schema.schemaId(), schema);
-      }
-
-      return builder.build();
-    }
-
     public ViewMetadata build() {
       Preconditions.checkArgument(null != location, "Invalid location: null");
       Preconditions.checkArgument(versions.size() > 0, "Invalid view: no versions were added");
 
-      checkCurrentVersionIdIsValid(currentVersionId);
-
-      int versionHistorySizeToKeep =
+      int historySize =
           PropertyUtil.propertyAsInt(
               properties,
               ViewProperties.VERSION_HISTORY_SIZE,
               ViewProperties.VERSION_HISTORY_SIZE_DEFAULT);
 
       Preconditions.checkArgument(
-          versionHistorySizeToKeep > 0,
+          historySize > 0,
           "%s must be positive but was %s",
           ViewProperties.VERSION_HISTORY_SIZE,
-          versionHistorySizeToKeep);
-
-      if (versions.size() > versionHistorySizeToKeep) {
-        List<ViewVersion> versionsToKeep =
-            versions.subList(versions.size() - versionHistorySizeToKeep, versions.size());
-        List<ViewHistoryEntry> historyToKeep =
-            history.subList(history.size() - versionHistorySizeToKeep, history.size());
-        List<MetadataUpdate> changesToKeep = Lists.newArrayList(changes);
-        Set<MetadataUpdate.AddViewVersion> toRemove =
-            changesToKeep.stream()
-                .filter(update -> update instanceof MetadataUpdate.AddViewVersion)
-                .map(update -> (MetadataUpdate.AddViewVersion) update)
-                .filter(
-                    update ->
-                        update.viewVersion().versionId() != currentVersionId
-                            && !versionsToKeep.contains(update.viewVersion()))
-                .collect(Collectors.toSet());
-        changesToKeep.removeAll(toRemove);
-
-        versions = versionsToKeep;
-        history = historyToKeep;
-        changes = changesToKeep;
+          historySize);
+
+      // expire old versions, but keep at least the versions added in this builder
+      int numAddedVersions = (int) changes(MetadataUpdate.AddViewVersion.class).count();
+      int numVersionsToKeep = Math.max(numAddedVersions, historySize);
+
+      List<ViewVersion> retainedVersions;
+      List<ViewHistoryEntry> retainedHistory;
+      if (versions.size() > numVersionsToKeep) {
+        retainedVersions = expireVersions(versionsById, numVersionsToKeep);
+        Set<Integer> retainedVersionIds =
+            retainedVersions.stream().map(ViewVersion::versionId).collect(Collectors.toSet());
+        retainedHistory = updateHistory(history, retainedVersionIds);
+      } else {
+        retainedVersions = versions;
+        retainedHistory = history;
       }
 
       return ImmutableViewMetadata.of(
           formatVersion,
           location,
-          schemas,
+          ImmutableList.copyOf(schemas),
```

Comment:
Using immutable collections now.

Reply:
no need to do this, as those will be immutable collections internally, which is being handled when the `ViewMetadata` instance is created